cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -92,7 +92,23 @@ type Etcd struct {
 	DataDir   string
 	ExtraArgs map[string]string
 	// Image specifies which container image to use for running etcd. If empty, automatically populated by kubeadm using the image repository and default etcd version
-	Image string
+	Image      string
+	SelfHosted *SelfHostedEtcd
+}
+
+// SelfHostedEtcd describes options required to configure self-hosted etcd
+type SelfHostedEtcd struct {
+	// CertificatesDir represents the directory where all etcd TLS assets are stored. By default this is
+	// a dir names "etcd" in the mains CertificatesDir value.
+	CertificatesDir string
+	// The service IP of the etcd cluster
+	ClusterIP string
+	// The number of members that will exist in the etcd cluster.
+	ClusterSize int
```

Comment:
Defining this in advance is somehow in contrast with the "dynamic flow" we are discussing in the HA proposal; basically the assumption here is that the number of etcd instances should be linked (and equal) to the number of master nodes, start with 1, and then increase automatically when we do `kubeadm join --master`

@luxas opinions here?

Reply:
Great catch on this, I agree. The operator also only ever schedules 1 pod per master too, so having this field is not necessary.